cfpb/consumerfinance.gov
# Stub Census API in functional tests

Our functional tests for the [rural and underserved areas tool](https://www.consumerfinance.gov/rural-or-underserved-tool/) rely on a couple of Census APIs to work. That's caused problems recently: the Census APIs are occasionally slow, which causes these functional tests to fail on pull requests that have nothing to do with the rural and underserved areas tool. We want to mock those requests instead so we're testing the functionality of the page instead of the speed of the Census API

---

## Additions

- Fixture for Census API response data
- Cypress intercepts for the Census API calls

## Changes

- Minor changes to test descriptions to make them more accurate

## How to test this PR

1.  Start Cypress with `yarn run cypress open --config blockHosts="*.census.gov"`. That config option will block requests to any census.gov URL from the Cypress test environment so you can be sure the intercepts are working.
2. Run the tests for the rural and underserved areas tool (integration tests > consumer-tools > rural-or-underserved-tool.js).
3. Verify that the tests pass and that all the calls to the Census API are intercepted and stubbed (see screenshot below)

## Screenshots

![ruat-tests](https://user-images.githubusercontent.com/1862695/156837802-553c8a1f-1f08-4002-9c46-d1d52bf2651d.png)

## Notes and todos

- The tool needs to make four different calls to two different Census APIs to get all the data it needs, and the API responses come back as JSONP, none of which Cypress handles super well. So I arranged things in the least ugly way I could think of, but I'm happy to rework that if anyone has a better way in mind.

## Checklist

  - PR titles are used to generate the change log in [releases](../../releases); good ones make that easier to scan.
  - Consider prefixing, e.g., "Mega Menu: fix layout bug", or "Docs: Update Docker installation instructions".
  - [This repo’s docs](https://cfpb.github.io/consumerfinance.gov/) (edit the files in the `/docs` folder) – for basic, close-to-the-code docs on working with this repo
  - CFGOV/platform wiki on GHE – for internal CFPB developer guidance
  - CFPB/hubcap wiki on GHE – for internal CFPB design and content guidance
